### ISSUE TITLE
Brainstorm on how to add custom op authoring

### ISSUE THREAD

**Issue by ramiro050** [llvm/torch-mlir#462]
- [Brainstorm on how to add custom op authoring](#brainstorm-on-how-to-add-custom-op-authoring)
- [Adding ops corresponding to lower-level ops](#adding-ops-corresponding-to-lower-level-ops)
  - [Going form TorchScript IR to Torch MLIR](#going-form-torchscript-ir-to-torch-mlir)
- [Inserting MLIR at different points in the lowering process](#inserting-mlir-at-different-points-in-the-lowering-process)

## Brainstorm on how to add custom op authoring

PyTorch offers many ops for performing computations. However, sometimes users want to have more fine-grained control over definitions of custom computations for which the building blocks from PyTorch are not enough. The purpose of this issue is to brainstorm on potential solutions for the problem of giving users a way to author custom ops at the PyTorch level without having to resort to writing C++ in `torch-mlir` or in PyTorch.

## Adding ops corresponding to lower-level ops

One approach that is straight-forward to implement is to add at the `torch` dialect level extra ops corresponding to lower level ops that we want to expose to users. For example, one can expose the `linalg.matmul` op by creating an op `torch.linalg.matmul`. This `torch` dialect op would then have a pretty simple lowering to `linalg`, corresponding to replacing `torch.linalg.matmul` with `linalg.matmul`. The user would then be able to write something like:

```python
class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x):
        z = torch.tanh(x)
        t = linalg.init_tensor([3,3], torch.float)
        return linalg.matmul(z, z, t)
```

This would become the following `torch` MLIR:

```mlir
module {
  func private @__torch__.MyModule.forward(%arg0: !torch.nn.Module<"__torch__.MyModule">, %arg1: !torch.tensor {torch.type_bound = !torch.vtensor<[3,3],f32>}) -> !torch.tensor {
    %int6 = torch.constant.int 6
    %int3 = torch.constant.int 3
    %1 = torch.aten.tanh %arg1 : !torch.tensor -> !torch.tensor
    %2 = torch.prim.ListConstruct %int3, %int3 : (!torch.int, !torch.int) -> !torch.list<!torch.int>
    %3 = torch.linalg.init_tensor %2, %int6 : !torch.list<!torch.int>, !torch.int -> !torch.tensor
    %4 = torch.linalg.matmul %1, %1, %3 : !torch.tensor, !torch.tensor, !torch.tensor -> !torch.tensor
    return %4 : !torch.tensor
  }
  
  // torch.class_type @__torch__.MyModule
  // %0 = torch.nn_module
}
```

This would then lower to the following `linalg` MLIR:

```mlir
#map = affine_map<(d0, d1) -> (d0, d1)>
module {
  func @forward(%arg0: tensor<3x3xf32>) -> tensor<3x3xf32> {
    %c3_i64 = arith.constant 3 : i64
    %0 = linalg.init_tensor [3, 3] : tensor<3x3xf32>
    %1 = linalg.generic {indexing_maps = [#map, #map], iterator_types = ["parallel", "parallel"]} ins(%arg0 : tensor<3x3xf32>) outs(%0 : tensor<3x3xf32>) {
    ^bb0(%arg1: f32, %arg2: f32):  // no predecessors
      %7 = math.tanh %arg1 : f32
      linalg.yield %7 : f32
    } -> tensor<3x3xf32>
    %2 = arith.index_cast %c3_i64 : i64 to index
    %3 = arith.index_cast %c3_i64 : i64 to index
    %4 = linalg.init_tensor [%2, %3] : tensor<?x?xf32>
    %5 = tensor.cast %4 : tensor<?x?xf32> to tensor<3x3xf32>
    %6 = linalg.matmul ins(%1, %1 : tensor<3x3xf32>, tensor<3x3xf32>) outs(%5 : tensor<3x3xf32>) -> tensor<3x3xf32>
    return %6 : tensor<3x3xf32>
  }
}
```

Main advantage: This is simple and can easily be code-generated, since most of the lowerings are just replacing an op from `torch` dialect with its corresponding op in the lower-level dialect.

Main disadvantage: This involves adding ops to the `torch` dialect that have nothing to do with PyTorch. This disadvantage makes it hard to justify this approach.

### Going form TorchScript IR to Torch MLIR

There are a couple of ways that I can think of for giving users at the PyTorch level access to ops from lower-level dialects like `linalg`. The first approach is to define a python module with definitions of the following form:

```python
# file: linalg.py
@torch.jit.ignore
def init_tensor(sizes: List[int], dtype: int) -> torch.Tensor:
    return


@torch.jit.ignore
def matmul(m1: torch.Tensor, m2: torch.Tensor, out: torch.Tensor
           ) -> torch.Tensor:
    return
```

Using `@torch.jit.ignore` allows users to compile their module using the TorchScript compiler, and the compiler will create a node with the name of the `linalg` op whenever it is used. Then, by making some simple changes to the `torch-mlir` node importer, one can identify the `torch.jit.ignore`'d functions, since they will be of `node->kind() == at::prim::PythonOp`, and then import them into the torch dialect.

Main advantage: This does not require changing PyTorch and it is very simple to create a module with ops for the user.

The second approach is to register the ops we want to expose to the user as PyTorch ops, which will require adding code to PyTorch.

Main advantage: This will not require changes to the `torch-mlir` importer.

Main disadvantage: This requires the user to build a special version of PyTorch.

## Inserting MLIR at different points in the lowering process

Another approach that I have thought less about is to give users Python bindings for functions that generate MLIR at a lower-level dialect like `linalg`. We can then capture the MLIR and insert it into the MLIR module being compiled after the `torch` ops have also been lowered to `linalg`. How to achieve this is not yet clear to me, especially when it comes to capturing the `linalg` MLIR while we wait for everything else to be lowered. This will also probably require a special op at the `torch` dialect level that will be a place-holder for a custom computation to be added later in the lowering process.

Main advantage: This allows us to re-use the already existing Python libraries for generating MLIR, such as the Python DSL for `linalg`.

Main disadvantage: It is unclear to me how to achieve this, and how much work it will require.

-------

I would love any feedback/ideas!

**Comment by silvasean** [comment]
What are the use cases we want to solve with this? The form of the solution will depend on that.

For the use case of squirreling some linalg ops through, I would prefer to register a Torch operator `torch_mlir::linalg_on_tensors_subroutine` and then attach a string of textual mlir in the linalg dialect to that.

**Comment by ds1231h** [comment]
> 

Thank you very much for this topic, and this is what I'm currently working on!

I'm trying the second way you mentioned here:
> Inserting MLIR at different points in the lowering process

For now, I'm planning to support an "extern_op" function in pytorch, like:
```
@torch.jit.ignore
def extern_op(self, in1, in2):
    return xxx
```
and convert that into the `library_call = "extern_op"` in linalg.

However, I met the problem that I don't know where to add such "extern_op" in `torch_ods_gen.py`, which I imagine like this(I want to implement the elementwise version first):
```
emit("prim::external_op : (Tensor) -> (Tensor)")
```

> What are the use cases we want to solve with this? 
I would like to use this method to call some efficient implementation from libs like Cudnn, CuBlas. Or we can extract the op that hasn't been implemented yet, and then users can implement that by themselves. What do you think?

> For the use case of squirreling some linalg ops through, I would prefer to register a Torch operator torch_mlir::linalg_on_tensors_subroutine and then attach a string of textual mlir in the linalg dialect to that.

Sorry, I didn't get that... Could you please explain it a bit? Thank you!

**Comment by silvasean** [comment]
> I would like to use this method to call some efficient implementation from libs like Cudnn, CuBlas. Or we can extract the op that hasn't been implemented yet, and then users can implement that by themselves. What do you think?

Torch-MLIR is not yet having a stable end-to-end flow that has high performance (IREE will probably be the first). So I think it is a bit early to consider libraries like Cudnn (at least in upstream on github). I think at the right time we can look into it though, but we should consider all approaches (such as linalg autotuning) to get higher performance.


> > For the use case of squirreling some linalg ops through, I would prefer to register a Torch operator torch_mlir::linalg_on_tensors_subroutine and then attach a string of textual mlir in the linalg dialect to that.
> 
> Sorry, I didn't get that... Could you please explain it a bit? Thank you!

We would use a custom Torch operator `torch_mlir::linalg_on_tensors_subroutine` with a string attached. The string would contain textual .mlir with the linalg dialect.

**Comment by powderluv** [comment]
> > I would like to use this method to call some efficient implementation from libs like Cudnn, CuBlas. Or we can extract the op that hasn't been implemented yet, and then users can implement that by themselves. What do you think?
> 
> Torch-MLIR is not yet having a stable end-to-end flow that has high performance (IREE will probably be the first). So I think it is a bit early to consider libraries like Cudnn (at least in upstream on github -- I encourage experiements if you want to do them). I think at the right time we can look into it though, but we should consider all approaches (such as linalg autotuning) to get higher performance.

I think cudnn is just an example. We do have call outs to hand optimized kernels at the lower levels downstream.  I agree upstream may lag a bit but there maturity of the downstream projects may not be evident right now (or ever ).  We are helping customers with custom_ops in torch-mlir so I'd be happy to convey any requirements that you may need from them.

**Comment by silvasean** [comment]
So, the current state is that if you register a Torch op in the c10 registry, we will import it opaquely as "torch.operator" op in MLIR and that will just pass through. Shape refinement and value semantics conversions won't work on such an opaque op, but if we then add it to ODS we can add annotations at least for value semantics. Shape/dtype propagation would be harder.

**Comment by ds1231h** [comment]
> Torch-MLIR is not yet having a stable end-to-end flow that has high performance (IREE will probably be the first). So I think it is a bit early to consider libraries like Cudnn (at least in upstream on github -- I encourage experiements if you want to do them). I think at the right time we can look into it though, but we should consider all approaches (such as linalg autotuning) to get higher performance.

Yeah, and the linalg autotuning sounds amazing, looking forward to it :)

> but if we then add it to ODS we can add annotations at least for value semantics.

That makes sense to me. 

I got the `%2 = torch.operator "prim.PythonOp"(%arg0, %0, %1) : (!torch.nn.Module<"__torch__.ExternCallModule">, !torch.tensor<[?,?],f32>, !torch.tensor<[?,?],f32>) -> !torch.tensor` in the IR before `TrochToLinalg`, and I can match it with `class ConvertOperatorOp : public OpConversionPattern<OperatorOp>` in my code. But I'm not clear how to add that to ODS. Do you mean adding a new function like `def emit_operator_ops(torch_ir_dir: str, registry: Registry):` in `torch_ods_gen.py`? I found the `TorchOps.td` isn't created automatically.

So I modified the `TorchOps.td`manually, but it doesn't work... 
Any advice please?

**Comment by silvasean** [comment]
@ds1231h for what you are doing, I would advise to define a custom op in TorchOps.td (not autogenerated), and then add code in node_importer.cpp to specially detect your op. That isn't a good design for upstream, but it should get you unblocked.

**Comment by ds1231h** [comment]
> @ds1231h for what you are doing, I would advise to define a custom op in TorchOps.td (not autogenerated), and then add code in node_importer.cpp to specially detect your op. That isn't a good design for upstream, but it should get you unblocked.

Thanks!

**Comment by silvasean** [comment]
Superceded by https://github.com/llvm/torch-mlir/issues/1462
